---
title: "BI586_dada2_analysis"
author: "Elsa Brenner, Jamie Poirier, Erica Sun"
date: "March 9, 2021"
output: html_document
---
VERSION CONTROL: 
```{r}
R.version
packageVersion("dada2") #1.16.0
packageVersion("ggplot2") #3.3.3
packageVersion("ShortRead") #1.46.0
packageVersion("phyloseq") #1.32.0
```
---
# **INTRODUCTION:** 

For this assignment, we used the data used in ["Patterns of environmental variability influence coral‐associated bacterial and algal communities on the Mesoamerican Barrier Reef"](https://onlinelibrary.wiley.com/doi/full/10.1111/mec.15497) published in *Molecular Ecology* by L. Speare, et. al. (2020).  

Speare et. al.'s study explores the bacterial and algal symbionts associated with two species of Caribbean stony corals-- *Siderastrea siderea* and *Siderastrea radians* found on the Belize Mesoamerican Barrier Reef (Figure 1).  Their study aimed to investigate the factors that determine variability in Bacterial Community Composition (BCC); specifically how BCC varried across two species in the same genus, and how BCC varried across location, both in different sites, and inshore v. offshore at specific sites.  

![*Siderastrea siderea*](/Users/elsa/Desktop/BI_586/BI586-Project1/BI586-Project1/Sid_photo.jpeg)
Figure 1: photo via *Coralpedia* (The University of Warwick)

For their analysis, Speare et. al. amplified a 444 bp section of the V3-V4 variable regions of the 16S gene. The 16S rRNA gene is commonly used in DNA barcoding of bacteria, and was therefore an optimal choice for investigating coral BCC.  

In brief summary, the team found that BCC varried significantly between species, and from site-to-site. However, their findings explained >11% of variability, and future investigation into the factors determining BCC and algal composition are necessary (Speare et. al., 2020).  

For this assignment, we looked at subsample of Speare et. al.'s data, 10 *Siderastrea siderea*, 5 from inshore locations and 5 from offshore locations, and accessed the genomic data from the [NCBI Sequence Read Archive (SRA)](https://www.ncbi.nlm.nih.gov/sra).  
---
# **STEP-BY-STEP METHODS AND DATA ANALYSIS:** 

## Getting Ready

### Downloading .fastq files from NCBI SRA
First, we have to download our data from [NCBI BioSample](https://www.ncbi.nlm.nih.gov/biosample?LinkName=bioproject_biosample_all&from_uid=628692). We chose 5 offshore and 5 inshore samples of *Siderastrea siderea* for our analysis. Now we can download all the sample number (starting with SRR) directly on NCBI by using the "send to" feature with settings:
Format = "Accessions List" and
Sort By = "Default order". 
Then save the accessions list with file name srafile.

Now we are ready to download our sequences.
```{css,echo=F}
.sccCode {
  background-color: black;
  color: white;}
```

```{bash, eval = F, echo = T, class.source="sccCode"}
$ prefetch --option-file srafile
```
We just downloaded all the sample runs as .sra files! Now we have to convert them into .fastq
```{bash, eval = F, echo = T, class.source="sccCode"}
# First let's make a directory to store our output .fastq files
$ mkdir sraout
# Convert them to .fastq and store the output in the sraout director
$ fastq-dump --split-files ./SRR* -O sraout/
```
That's it! Now we have all of our sampels in .fastq format and we're ready to run the dada2 pipeline!

### Loading Required Packages
First we have to install the packages. 
```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("dada2")
BiocManager::install('vegan')
BiocManager::install('phyloseq')
```
Then load the required packages.
```{r, eval=FALSE}
library(dada2)
library(ShortRead)
library(ggplot2)
library(phyloseq)
```

### Set the working directory
We need to make sure that all the files we are working with are in the same directory.
```{r}
# Make sure to change this to the directory you're working in!
setwd('/Users/ericasun/Documents/GitHub/BI586-Project1/')
```

## Trimming and Filtering
Set our current working directory as path, and set all of the files in there as fns. 
```{r}
path <- '/Users/ericasun/Documents/GitHub/BI586-Project1/'
fns <- list.files(path)
```

Align every file that ends with .fastq as fastqs and sort them.
```{r}
fastqs <- fns[grepl(".fastq$", fns)]
fastqs <- sort(fastqs) 
# Now let's take a look to ensure we're on the right track!
fastqs
```

Get the sample names, so essentailly just everything before "_1.fastq"
```{r}
sample.names <- sapply(strsplit(fastqs, "_1.fastq"), `[`, 1)
# Let's take a look
sample.names
```
Assign the full path as fnFs
```{r}
fnFs <- file.path(path, fastqs)
head(fnFs) #Let's just take a peak this time
```

## Visualize Raw Data
Let's take a look at quality profile
```{r}
# Let's start with the first five reads, which are the offshores. 
dada2::plotQualityProfile(fnFs[c(1,2,3,4,5)])
# Then let's look at the inshore ones
dada2::plotQualityProfile(fnFs[c(6,7,8,9,10)])
```
---
# **REFERENCES:**

Coralpedia - Your guidie to caribbean corals and sponges. (2021). Retrieved March 05, 2021, from https://coralpedia.bio.warwick.ac.uk/en/corals/siderastrea_siderea

Speare, L, Davies, SW, Balmonte, JP, Baumann, J, Castillo, KD. Patterns of environmental variability influence coral‐associated bacterial and algal communities on the Mesoamerican Barrier Reef. Mol Ecol. 2020; 29: 2334– 2348. https://doi.org/10.1111/mec.15497

[WoRMS Editorial Board](http://www.marinespecies.org/aphia.php?p=popup&name=citation) (2021). World Register of Marine Species. Available from http://www.marinespecies.org at VLIZ. Accessed 2021-03-05. doi:10.14284/170