---
title: "BI586_dada2_analysis"
author: "Elsa Brenner, Jamie Poirier, Erica Sun"
date: "March 9, 2021"
output: 
  html_document:
    toc: TRUE
    toc_depth: 3
    toc_float: TRUE
    theme: lumen
---
VERSION CONTROL: 
```{r}
R.version
packageVersion("dada2") #1.16.0
packageVersion("ggplot2") #3.3.3
packageVersion("ShortRead") #1.46.0
packageVersion("phyloseq") #1.32.0
```
---
# **INTRODUCTION:** 

For this assignment, we used the data used in ["Patterns of environmental variability influence coral‐associated bacterial and algal communities on the Mesoamerican Barrier Reef"](https://onlinelibrary.wiley.com/doi/full/10.1111/mec.15497) published in *Molecular Ecology* by L. Speare, et. al. (2020).  

Speare et. al.'s study explores the bacterial and algal symbionts associated with two species of Caribbean stony corals-- *Siderastrea siderea* and *Siderastrea radians* found on the Belize Mesoamerican Barrier Reef (Figure 1).  Their study aimed to investigate the factors that determine variability in Bacterial Community Composition (BCC); specifically how BCC varried across two species in the same genus, and how BCC varried across location, both in different sites, and inshore v. offshore at specific sites.  

![*Siderastrea siderea*](/Users/elsa/Desktop/BI_586/BI586-Project1/BI586-Project1/Sid_photo.jpeg)
Figure 1: photo via *Coralpedia* (The University of Warwick)

For their analysis, Speare et. al. amplified a 444 bp section of the V3-V4 variable regions of the 16S gene. The 16S rRNA gene is commonly used in DNA barcoding of bacteria, and was therefore an optimal choice for investigating coral BCC.  

In brief summary, the team found that BCC varried significantly between species, and from site-to-site. However, their findings explained >11% of variability, and future investigation into the factors determining BCC and algal composition are necessary (Speare et. al., 2020).  

For this assignment, we looked at subsample of Speare et. al.'s data, 10 *Siderastrea siderea*, 5 from inshore locations and 5 from offshore locations, and accessed the genomic data from the [NCBI Sequence Read Archive (SRA)](https://www.ncbi.nlm.nih.gov/sra).  
---
# **STEP-BY-STEP METHODS AND DATA ANALYSIS:** 

## Getting Ready

### Downloading .fastq files from NCBI SRA
First, we have to download our data from [NCBI BioSample](https://www.ncbi.nlm.nih.gov/biosample?LinkName=bioproject_biosample_all&from_uid=628692). We chose 5 offshore and 5 inshore samples of *Siderastrea siderea* for our analysis. Now we can download all the sample number (starting with SRR) directly on NCBI by using the "send to" feature with settings:
Format = "Accessions List" and
Sort By = "Default order". 
Then save the accessions list with file name srafile.

Now we are ready to download our sequences.
```{css,echo=F}
.sccCode {
  background-color: black;
  color: white;}
```

```{bash, eval = F, echo = T, class.source="sccCode"}
$ prefetch --option-file srafile
```
We just downloaded all the sample runs as .sra files! Now we have to convert them into .fastq
```{bash, eval = F, echo = T, class.source="sccCode"}
# First let's make a directory to store our output .fastq files
$ mkdir sraout
# Convert them to .fastq and store the output in the sraout director
$ fastq-dump --split-files ./SRR* -O sraout/
```
That's it! Now we have all of our sampels in .fastq format and we're ready to run the dada2 pipeline!

### Loading Required Packages
First we have to install the packages. 
```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("dada2")
BiocManager::install('vegan')
BiocManager::install('phyloseq')
```
Then load the required packages.
```{r, eval=FALSE}
library(dada2)
library(ShortRead)
library(ggplot2)
library(phyloseq)
```

### Set the working directory
We need to make sure that all the files we are working with are in the same directory.
```{r}
# Make sure to change this to the directory you're working in!
setwd('/Users/ericasun/Documents/GitHub/BI586-Project1/')
```

## Trimming and Filtering
Set our current working directory as path, and set all of the files in there as fns. 
```{r}
path <- '/Users/ericasun/Documents/GitHub/BI586-Project1/'
fns <- list.files(path)
```

Align every file that ends with .fastq as fastqs and sort them.
```{r}
fastqs <- fns[grepl(".fastq$", fns)]
fastqs <- sort(fastqs) 
# Now let's take a look to ensure we're on the right track!
fastqs
```

Get the sample names, so essentailly just everything before "_1.fastq"
```{r}
sample.names <- sapply(strsplit(fastqs, "_1.fastq"), `[`, 1)
# Let's take a look
sample.names
```
Assign the full path as fnFs
```{r}
fnFs <- file.path(path, fastqs)
head(fnFs) #Let's just take a peak this time
```

### Visualize Raw Data
Let's take a look at quality profile
```{r}
# Let's start with the first five reads, which are the offshores. 
dada2::plotQualityProfile(fnFs[c(1,2,3,4,5)])
# Then let's look at the inshore ones
dada2::plotQualityProfile(fnFs[c(6,7,8,9,10)])
```
According to the output, we should trim our data at 200. Let's create a directory and filenames of the trimmed fastqs. 
```{r}
filt_path <- file.path(path, "trimmed")
if(!file_test("-d", filt_path)) dir.create(filt_path)
filtFs <- file.path(filt_path, paste0(sample.names, "_filt.fastq.gz"))
```

### Filtering
```{r}
out <- dada2::filterAndTrim(fnFs, filtFs, truncLen= 200, #end of single end reads = approx. 300 bp
                     maxN=0, #DADA does not allow Ns
                     maxEE=1, #allow 1 expected errors, where EE = sum(10^(-Q/10)); more conservative, model converges
                     truncQ=2, #standard filtering parameters
                     trimLeft=20, #N nucleotides to remove from the start of each read: ITS2 primer = F 20bp
                     rm.phix=TRUE, #remove reads matching phiX genome
                     compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE

head(out)
tail(out)
```
## Learn Error Rates
Every amplicon dataset has a different set of error rates, the learnError method can learn the error model from the data. The algorithm must begin with an initial guess, for which the maximum possible error rates in the data are used. For example, when the most abundant sequence is correct, and all the rest are errors.  
Some things to keep in mind --  
* Error rates should decline with increasing qual score
* red line is based on definition of quality score alone
* black line is estimated error rate after convergence
```{r}
#setDadaOpt(MAX_CONSIST=30) #increase number of cycles to allow convergence #Actually not sure what this is doing? Ask for help.
errF <- dada2::learnErrors(filtFs, multithread=TRUE)
dada2::plotErrors(errF, nominalQ=TRUE) 
```

## Dereplicate Reads
Dereplication combines identical sequences into one "unique sequence". "Abundance" is the number of reads with that unique sequence. This is done to reduce computation time by eliminating redundant comparisons. DADA2 has a summary of the quality information associated with each unique sequence. To increase DADA2's accuracy, this quality information tells the error model of the subsequent denoising step.
```{r}
derepFs <- dada2::derepFastq(filtFs, verbose=TRUE)
# Naming the dereplicated-class objects by the sample names
names(derepFs) <- sample.names
```

## Infer Sequence Variants
```{r}
dadaFs <- dada2::dada(derepFs, err=errF, multithread=TRUE)
```
Now, let's take a look at the data class object by sample. This will tell us how many "real" variants there are in the unique input seqs.
```{r}
dadaFs[[1]]
dadaFs[[10]]
```
The DADA2 algorithm inferred 144 true sequence variants from the 4463 unique sequences in the first sample. 
The DADA2 algorithm inferred 80 true sequence variants from the 1785 unique sequences in the first sample. 
Now let's construct a sequnce table.
```{r, results='hide'}
seqtab <- dada2::makeSequenceTable(dadaFs)
head(seqtab)
```

## Remove Chimeras
We have removed the substitution and idel errors, but there are still chimeras. So we're going to remove them in this step. The accuracy of the sequences after denoisying makes identifying chimeras easier. All sequences can be exactly reconstructed as a bimera (two-parent chimera) from more abundant sequences. 
```{r}
seqtab.nochim <- dada2::removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim) # dim function retrieve or set the dimension of an object
```
We identified 6 bimeras out of 940 input sequences. This is how much sample remain after the removal of chimera. Most reads still remain. 
```{r}
sum(seqtab.nochim)/sum(seqtab)
```
For our sample, the ration was 0.9956432 with 6 bimeras. Let's output this result as a .csv file
```{r}
write.csv(seqtab,file="dadaFs_seqtab.csv")
write.csv(seqtab.nochim,file="dadaFs_nochim.csv")
```

## Track Read Stats
```{r}
getN <- function(x) sum(dada2::getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaFs, getN), rowSums(seqtab), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoised", "merged", "tabled", "nonchim")
rownames(track) <- sample.names
head(track)
tail(track)

write.csv(track,file="ReadFilterStats_AllData_final.csv",row.names=TRUE,quote=FALSE)
```

## Assign Taxonomy
DADA2 uses naive Bayesian classifier method to assign taxonomy. The assignTaxonomy function takes as input a set of sequences to be classified and a tranning set of reference sequence with known taxonomy. It outputs taxonomic assignments with at least minBoot bootstrap confidence.
```{r}
taxa <- dada2::assignTaxonomy(seqtab.nochim, "GeoSymbio_ITS2_LocalDatabase_verForPhyloseq.fasta", minBoot=70,multithread=TRUE,tryRC=TRUE,outputBootstraps=FALSE)
#minboot should be 70
#Obtain a csv file for the taxonomy so that it's easier to map the sequences for the heatmap.
write.csv(taxa, file="taxa.csv",row.name=TRUE,quote=FALSE)
unname(head(taxa, 30))
unname(taxa)

# Save outputs as .rds to save some spaces
saveRDS(seqtab.nochim, file="final_seqtab_nochim.rds")
saveRDS(taxa, file="final_taxa_blastCorrected.rds")

# If you need to read in previously saved datafiles
seqtab.nochim <- readRDS("final_seqtab_nochim.rds")
taxa <- readRDS("final_taxa_blastCorrected.rds")
head(taxa)
```

## Handoff to Phyloseq
Pyloseq is going to give us some visualization! Import the required libraries.
```{r}
library(phyloseq)
library(ggplot2)
```

---
# **REFERENCES:**

Coralpedia - Your guidie to caribbean corals and sponges. (2021). Retrieved March 05, 2021, from https://coralpedia.bio.warwick.ac.uk/en/corals/siderastrea_siderea

Speare, L, Davies, SW, Balmonte, JP, Baumann, J, Castillo, KD. Patterns of environmental variability influence coral‐associated bacterial and algal communities on the Mesoamerican Barrier Reef. Mol Ecol. 2020; 29: 2334– 2348. https://doi.org/10.1111/mec.15497

[WoRMS Editorial Board](http://www.marinespecies.org/aphia.php?p=popup&name=citation) (2021). World Register of Marine Species. Available from http://www.marinespecies.org at VLIZ. Accessed 2021-03-05. doi:10.14284/170


